File: packages/core/src/programmers/json/JsonSchemaProgrammer.ts
import { IJsonSchemaUnit, OpenApi } from "@typia/interface";
import ts from "typescript";

import { ITypiaContext } from "../../context/ITypiaContext";
import { LiteralFactory } from "../../factories/LiteralFactory";
import { MetadataSchema } from "../../schemas/metadata/MetadataSchema";
import { JsonSchemasProgrammer } from "./JsonSchemasProgrammer";

export namespace JsonSchemaProgrammer {
  export const validate = (metadata: MetadataSchema): string[] =>
    JsonSchemasProgrammer.validate(metadata);

  export interface IWriteProps<Version extends "3.0" | "3.1"> {
    context: ITypiaContext;
    version: Version;
    metadata: MetadataSchema;
  }

  export const write = <Version extends "3.0" | "3.1">(
    props: IWriteProps<Version>,
  ): ts.Expression => {
    const schema: IJsonSchemaUnit<Version> = writeSchema({
      version: props.version,
      metadata: props.metadata,
    });

    return ts.factory.createAsExpression(
      LiteralFactory.write(schema),
      props.context.importer.type({
        file: "typia",
        name: "IJsonSchemaUnit",
        arguments: [
          ts.factory.createLiteralTypeNode(
            ts.factory.createStringLiteral(props.version),
          ),
        ],
      }),
    );
  };

  export const writeSchema = <Version extends "3.0" | "3.1">(props: {
    version: Version;
    metadata: MetadataSchema;
  }): IJsonSchemaUnit<Version> => {
    const collection = JsonSchemasProgrammer.writeSchemas({
      version: props.version,
      metadatas: [props.metadata],
    });
    return {
      version: collection.version as "3.1",
      components: collection.components as OpenApi.IComponents,
      schema: collection.schemas[0] as OpenApi.IJsonSchema,
    } as IJsonSchemaUnit<Version>;
  };
}

File: packages/typia/native/core/programmers/json/JsonSchemaProgrammer.go
package json

import (
	"strings"

	shimast "github.com/microsoft/typescript-go/shim/ast"
	nativecontext "github.com/samchon/typia/packages/typia/native/core/context"
	nativefactories "github.com/samchon/typia/packages/typia/native/core/factories"
	nativeiterate "github.com/samchon/typia/packages/typia/native/core/programmers/iterate"
	nativemetadata "github.com/samchon/typia/packages/typia/native/core/schemas/metadata"
)

type jsonSchemaProgrammerNamespace struct{}

var JsonSchemaProgrammer = jsonSchemaProgrammerNamespace{}

type JsonSchemaProgrammer_IWriteProps struct {
	Context  nativecontext.ITypiaContext
	Version  string
	Metadata *nativemetadata.MetadataSchema
}

type JsonSchemaProgrammer_IJsonSchemaUnit struct {
	Version    string
	Components *nativeiterate.OpenApi_IComponents
	Schema     nativeiterate.JsonSchema
}

func (jsonSchemaProgrammerNamespace) Validate(props struct {
	Metadata *nativemetadata.MetadataSchema
	Explore  nativefactories.MetadataFactory_IExplore
}) []string {
	return JsonSchemasProgrammer.Validate(props)
}

func (jsonSchemaProgrammerNamespace) Write(props JsonSchemaProgrammer_IWriteProps) *shimast.Node {
	schema := JsonSchemaProgrammer.WriteSchema(struct {
		Version  string
		Metadata *nativemetadata.MetadataSchema
	}{
		Version:  props.Version,
		Metadata: props.Metadata,
	})
	return nativefactories.LiteralFactory.Write(map[string]any{
		"version":    schema.Version,
		"components": schema.Components,
		"schema":     schema.Schema,
	})
}

func (jsonSchemaProgrammerNamespace) WriteSchema(props struct {
	Version  string
	Metadata *nativemetadata.MetadataSchema
}) JsonSchemaProgrammer_IJsonSchemaUnit {
	collection := JsonSchemasProgrammer.WriteSchemas(struct {
		Version   string
		Metadatas []*nativemetadata.MetadataSchema
	}{
		Version:   props.Version,
		Metadatas: []*nativemetadata.MetadataSchema{props.Metadata},
	})
	var schema nativeiterate.JsonSchema
	if len(collection.Schemas) != 0 {
		schema = jsonSchemaProgrammer_dereference_root(collection.Schemas[0], collection.Components)
	}
	return JsonSchemaProgrammer_IJsonSchemaUnit{
		Version:    collection.Version,
		Components: collection.Components,
		Schema:     schema,
	}
}

func jsonSchemaProgrammer_dereference_root(schema nativeiterate.JsonSchema, components *nativeiterate.OpenApi_IComponents) nativeiterate.JsonSchema {
	if schema == nil || components == nil || components.Schemas == nil {
		return schema
	}
	ref, ok := schema["$ref"].(string)
	if ok == false {
		return schema
	}
	const prefix = "#/components/schemas/"
	if strings.HasPrefix(ref, prefix) == false {
		return schema
	}
	target, ok := components.Schemas[strings.TrimPrefix(ref, prefix)]
	if ok == false {
		return schema
	}
	output := nativeiterate.JsonSchema{}
	for key, value := range target {
		output[key] = value
	}
	return output
}
